README
$Id: README,v 1.2 2010/02/02 11:47:18 ak Exp $
                                            
 #####  ######   ##   ####   ##  ## ######  
 ##  ## ##      ####  ## ##  ###### ##      
 ##  ## ####   ##  ## ##  ## ###### ####    
 #####  ##     ###### ##  ## ##  ## ##      
 ## ##  ##     ##  ## ## ##  ##  ## ##      
 ##  ## ###### ##  ## ####   ##  ## ######  
                                            
--------------------------------------------------------------------------------
 sendmail-virtualmailbox: Virtual Mailbox(Maildir) Implementation with sendmail
--------------------------------------------------------------------------------
 sendmail-virtualmailbox(virtmail.m4) is a MAILER which provides virtual mailbox
 feature to sendmail. It require the following:

 PROGRAMS
 * sendmail: It is installed into /usr/sbin/sendmail
 * procmail: It delivers email to a local user(has UNIX account)'s Maildir/
 * maildrop: It delivers email to a virtual user(has NO UNIX account)'s Maildir/

 USERS AND GROUPS
 * sendmail: O DefaultUser=sendmail
 * smsubmit: Owner of /var/spool/cqueue (submission)
 * virtmail: Owner of virtual mailbox direcotry: --virtualmailbox-home=DIR
--------------------------------------------------------------------------------

CHANGES ------------------------------------ Version histoty
DISTFILES ---------------------------------- File list after installation
INSTALL ------------------------------------ How to install and configure virtual mailbox
KNOWNBGUS ---------------------------------- Known bugs in virtmail.m4
LICENSE ------------------------------------ This software's license
README ------------------------------------- This file
TODO --------------------------------------- Not implemented or tested yet
configure ---------------------------------- configure script for making sendmail-virtualmailbox
build/ ------------------------------------- Base directory for making procmail, maildrop, and sendamil
 maildrop-configure.sh --------------------- Configure script source for making maidrop
 procmail-conflict-types-for-getline.patch - Patch for the error 'Conflicting types for getline' 
 procmail-localconfig.diff ----------------- Patch for procmail to change paths.
 sendmail-mc.in ---------------------------- /usr/local/sendmail/etc/sendmail.cf source
 sendmail-site.config.m4 ------------------- Build macros for sendmail
 sid-milter-localconfig.diff --------------- Patch for sid-milter to change paths.
 submit-mc.in ------------------------------ /usr/local/sendmail/etc/submit.cf source
 virtmail.m4.in ---------------------------- source of cf/mailer/virtmail.m4
examples/ ---------------------------------- Some example configuration files
 INSTALL-MANUALLY -------------------------- ./INSTALL of v0.1.4 for manual installation
 local-sendmail/ --------------------------- Files for manual operations to setup /usr/local/sendmail
 usr-sbin-sendmail/ ------------------------ Files for manual operations to setup /usr/local/virtmail with /usr/sbin/sendmail
 a-part-of-aliases ------------------------- Example of /etc/mail/aliases
 a-part-of-dovecot.conf -------------------- Example of dovecot.conf (a part of) 
 clamav-configure -------------------------- Example of configure for clamav
 dovecot.conf-example ---------------------- dovecot.conf example
 dovecot-configure ------------------------- Example of configure for dovecot
 dovecot-pkgsysconfdir.patch --------------- /usr/local/dovecot/etc/dovecot -> /usr/local/dovecot/etc
 dovecot-setup ----------------------------- Script for dovecot after ``make install''
 dovecot-virtuser-passwd ------------------- Example of dovecot's passwd file
 etc-passwd -------------------------------- Required users for sendmail-virtualmailbox
 etc-group --------------------------------- Required groups for sendmail-virtualmailbox
src/ --------------------------------------- Source files
 etc/ -------------------------------------- Source files for /usr/local/sendmail/etc
  access.in -------------------------------- /etc/mail/access
  daemon-options.in ------------------------ /etc/mail/daemon-options
  etc-makefile.in -------------------------- /etc/mail/Makefile
  generics-domains.in ---------------------- /etc/mail/generics-domains
  genericstable.in ------------------------- /etc/mail/genericstable
  local-host-names.in ---------------------- /etc/mail/local-host-names
  maildroprc.in ---------------------------- /usr/local/virtmail/etc/maildroprc
  mailertable.in --------------------------- /etc/mail/mailertable
  procmailrc.in ---------------------------- /usr/local/virtmail/etc/procmailrc
  queue-interval.in ------------------------ /etc/mail/queue-interval
  relay-domains.in ------------------------- /etc/mail/relay-domains
  sid-options.in --------------------------- Define sid-milter options
  sid-pass-domains.in ---------------------- Domains to always pass for sid-milter
  sid-peer-hosts.in ------------------------ List of hosts to ignore for sid-milter
  virtuser-domains.in ---------------------- /etc/mail/virtuser-domains
  virtusertable.in ------------------------- /etc/mail/virtusertable
 sbin/ ------------------------------------- Source scripts for /usr/local/sendmail/sbin
  make-vmailbox.SH ------------------------- /usr/local/virtmail/sbin/make-vmailbox
  sendmailctl.SH --------------------------- /usr/local/virtmail/sbin/sendmailctl, /usr/local/sbin/smctl
  sidctl.SH -------------------------------- /usr/local/virtmail/sbin/sidctl

virtmail.m4 has been tested at the followings:
 sendmail) 8.13.8, 8.14.3 - 8.14.5
 dovecot)  1.2.9 - 1.2.16
 procmail) 3.22
 maildrop) 2.3.0.20100112, 2.5.1, 2.5.2, 2.5.3



CHANGES
# $Id: $

  ####  ##  ##   ##   ##  ##  ####  ######  #####  
 ##  ## ##  ##  ####  ### ## ##  ## ##     ###     
 ##     ###### ##  ## ###### ##     ####    ###    
 ##     ##  ## ###### ## ### ## ### ##       ###   
 ##  ## ##  ## ##  ## ##  ## ##  ## ##        ###  
  ####  ##  ## ##  ## ##  ##  ####  ###### #####   

2011/06/03	Directory name changed: mda -> virtmail
2011/05/19	sendmail 8.14.5
2011/04/27	maildrop 2.5.3
2011/03/30	maildrop 2.5.2Z
2010/11/23	Add procmail-conflict-types-for-getline.patch
2010/10/27	maildrop 2.5.1 and dovecot 1.2.15 .
2010/03/08	Add more example configuration files.
2010/02/08	Add more example configuration files.
2010/02/07	Fix ruleset 98/ParseLocal bug.
2010/02/06	Use git repository.
2009/01/01	cf/feature -> cf/mailer .
2008/07/11	Rename: virtualmailbox.m4 -> virtmail.m4
2008/04/29	Register into CVS repository.
2007/12/29	Test version, virtualmailbox.m4 only.



src/sbin/sidctl.SH
#!/bin/sh
# $Id: sidctl.SH,v 5.43 2008/02/28 06:16:16 ak Exp $
# @(#) sidctl(8) - The control interface of sid-milter
#      _     _      _   _ 
#  ___(_) __| | ___| |_| |
# / __| |/ _` |/ __| __| |
# \__ \ | (_| | (__| |_| |
# |___/_|\__,_|\___|\__|_|
#                         
readonly MYNAME="`basename $0`"

KILL_SIGNAL=15
DEBUG_LEVEL=''
SMLOG_LEVEL=''
RUN_MAKEMAP='0'
STDERRORDEV='/dev/stderr'
STATUS_CODE=1
CHAR_OK='[ OK ]'
CHAR_NG='[FAIL]'
CHAR_WARN='[WARN]'
CHAR_INFO='[INFO]'
CHAR_SKIP='[SKIP]'

# * Error or exit status code. see sysexits.h
EX_OK=0		# operation completed successfully
EX_GENERAL=1	# generic error
EX_USAGE=64	# invalid gesogram option or argument
EX_SOFTWARE=70	# internal software error
EX_OSERR=71	# system error
EX_OSFILE=72	# critical OS file missing
EX_CONFIG=78	# gesogram config | environment error
EX_NOPERM=77	# permission denied


sidcontrol()
{
	local whataction=$1
	local procidfile="/var/run/sid-milter.pid"
	local commandarg="/usr/local/bin/sid-filter"
	local daemonopts="`test -s /etc/mail/sid-options && grep -v '^#' /etc/mail/sid-options`"

	if [ "${whataction}X" = "startX" ]; then

		msg 'Starting sid-milter ... '
		if [ -s "$procidfile" ]; then
			msg ", $procidfile exists"
			result NG
			STATUS_CODE=$EX_GENERAL
		else
			commandarg="$commandarg -a /etc/mail/sid-peer-hosts"
			commandarg="$commandarg -d /etc/mail/sid-pass-domains"
			commandarg="$commandarg -P $procidfile"
			commandarg="$commandarg $daemonopts"
			( $commandarg && sleep 1 )

			if [ "$?" -eq "0" -a -s "$procidfile" ]; then
				echo "$commandarg" >> $procidfile
				result OK
				STATUS_CODE=$EX_OK
			else
				result NG
				STATUS_CODE=$EX_GENERAL
			fi
		fi

	elif [ "${whataction}X" = "stopX" ]; then

		msg 'Stopping sid-milter ... '
		if [ -s "$procidfile" ]; then
			( kill -$KILL_SIGNAL `head -1 $procidfile` && sleep 1 )
			if [ "$?" -eq "0" -a ! -s "$procidfile" ]; then
				result OK
				STATUS_CODE=$EX_OK
			else
				result NG
				STATUS_CODE=$EX_GENERAL
			fi
		else
			msg ", $procidfile does not exist"
			result NG
			STATUS_CODE=$EX_GENERAL
		fi
	fi
}

sidstatus()
{
	local procidfile="/var/run/sid-milter.pid"
	if [ -f "$procidfile" ]; then
		msg "sid-milter is running, pid = `head -n 1 $procidfile`"
		echo > $STDERRORDEV
		msg "command args = `tail -n 1 $procidfile`"
		echo > $STDERRORDEV
	fi
}

proclist()
{
	ps axuww | grep bin/sid-filter | grep -v grep | sed 's| *$||g'
}

msg()
{
	[ "${TERM}X" = "X" ] && STDERRORDEV='/dev/null'
	echo -n "$1" > $STDERRORDEV
}

result()
{
	local msgtype="$1"
	local message=''
	[ "${TERM}X" = "X" ] && STDERRORDEV='/dev/null'

	case "$msgtype" in
		ok|OK)
			message=$CHAR_OK
			;;

		ng|NG)
			message=$CHAR_NG
			;;

		warn|WARN)
			message=$CHAR_WARN
			;;

		info|INFO)
			message=$CHAR_INFO
			;;

		skip|SKIP)
			message=$CHAR_SKIP
			;;
		*)
			;;
	esac

	printf \\033[70G > $STDERRORDEV
	printf \\033[1\;49m > $STDERRORDEV
	echo -n $message > $STDERRORDEV
	printf \\033[0\;39m > $STDERRORDEV
	echo > $STDERRORDEV
}

init()
{
	export PATH='/bin:/usr/bin:/sbin:/usr/sbin:/usr/local/bin:/usr/local/sbin'
	export LANG='C'
	if [ -x "/usr/local/bin/sid-filter" ]; then 
		SIDFILTER="/usr/bin/sid-filter"
	else
		echo "Cannot execute /usr/local/bin/sid-filter" > $STDERRORDEV
		exit $EX_GENERAL
	fi
}

help()
{
	echo
	echo 'usage:'
	echo " $MYNAME [ OPTION [ARGV] ] COMMAND"
	echo
	echo '  COMMAND:'
	echo '     start        : Start sid-milter.'
	echo '     stop         : Stop sid-milter.'
	echo '     restart      : Restart ( stop & start )'
	echo '     status,s     : Print sid-milter status, content of pid file.'
	echo '     ps           : Print sid-milter processes.'
	echo '     help,h       : Print this screen. '
	echo 
}


#  ____ ____ ____ ____ 
# ||m |||a |||i |||n ||
# ||__|||__|||__|||__||
# |/__`|/__`|/__`|/__`|
#
init
case "$1" in
	#  ____ ____ ____ _________ ____ ____ ____ ____ ____ ____ ____ ____ 
	# ||s |||u |||b |||       |||c |||o |||m |||m |||a |||n |||d |||s ||
	# ||__|||__|||__|||_______|||__|||__|||__|||__|||__|||__|||__|||__||
	# |/__`|/__`|/__`|/_______`|/__`|/__`|/__`|/__`|/__`|/__`|/__`|/__`|
	#
	help|h)
		help
		exit $EX_OK
		;;
	ps)
		# Print processes.
		proclist
		exit $EX_OK
		;;
	restart)
		sidcontrol stop
		sidcontrol start
		;;

	start)
		sidcontrol start
		;;

	status)
		sidstatus
		;;
	stop|abort)
		if [ "$1" = "abort" ]; then
			KILL_SIGNAL=9	
		fi
		sidcontrol stop
		;;
	
	*)
		echo "try $MYNAME help"
		exit $EX_USAGE
		;;
esac




src/sbin/make-vmailbox.SH
#!/bin/sh
# $Id: make-vmailbox.SH,v 1.12.2.1 2011/03/19 09:41:17 ak Exp $
export LANG=C
export PATH="/bin:/usr/bin:/usr/local/bin:/sbin:/usr/sbin:/usr/local/sbin"
VIRTMAILUSER="__VIRTMAILUSER__"
VIRTMAILHOME="__VIRTMAILHOME__"
NEWMAILADDRS="$@"

if [ -n "$NEWMAILADDRS" ]; then
	test -d $VIRTMAILHOME || \
		( mkdir -pm 0755 $VIRTMAILHOME && chown $VIRTMAILUSER:$VIRTMAILUSER $VIRTMAILHOME )

	for E in $NEWMAILADDRS; do
		lpart="`echo $E | cut -d@ -f1`"
		dpart="`echo $E | cut -d@ -f2`"

		if [ -n "$lpart" -a -n "$dpart" ]; then
			if [ ! -d "$VIRTMAILHOME/$dpart/$lpart/Maildir" ]; then
				su - $VIRTMAILUSER -c "mkdir -pm 0700 $VIRTMAILHOME/$dpart/$lpart/Maildir/cur"
				su - $VIRTMAILUSER -c "mkdir -pm 0700 $VIRTMAILHOME/$dpart/$lpart/Maildir/new"
				su - $VIRTMAILUSER -c "mkdir -pm 0700 $VIRTMAILHOME/$dpart/$lpart/Maildir/tmp"
				echo "$E => $VIRTMAILHOME/$dpart/$lpart/Maildir"
			fi
		else
			echo "'$E' does not seem to email address"
		fi
	done
else
	echo "Usage: "
	echo " $0 email1 [ email2 [ email3 [...] ] ]"
	echo
	echo "Example: $0 hoge@example.jp fuga@example.org"
	exit 1;
fi


src/sbin/sendmailctl.SH
#!/bin/sh
# $Id: sendmailctl.SH,v 5.43 2008/02/28 06:16:16 ak Exp $
# @(#) sendmailctl(8) - The control interface of sendmail-virtualmailbox
#                     _                 _ _      _   _ 
#  ___  ___ _ __   __| |_ __ ___   __ _(_) | ___| |_| |
# / __|/ _ \ '_ \ / _` | '_ ` _ \ / _` | | |/ __| __| |
# \__ \  __/ | | | (_| | | | | | | (_| | | | (__| |_| |
# |___/\___|_| |_|\__,_|_| |_| |_|\__,_|_|_|\___|\__|_|
#
readonly SYSVERSION='__SYSVERSION__'
readonly SMUSER='__SENDMAILUSER__'
readonly MYNAME="`basename $0`"

KILL_SIGNAL=15
DEBUG_LEVEL=''
SMLOG_LEVEL=''
RUN_MAKEMAP='0'
STDERRORDEV='/dev/stderr'
STATUS_CODE=1
CHAR_OK='[ OK ]'
CHAR_NG='[FAIL]'
CHAR_WARN='[WARN]'
CHAR_INFO='[INFO]'
CHAR_SKIP='[SKIP]'

# * Error or exit status code. see sysexits.h
EX_OK=0		# operation completed successfully
EX_GENERAL=1	# generic error
EX_USAGE=64	# invalid gesogram option or argument
EX_SOFTWARE=70	# internal software error
EX_OSERR=71	# system error
EX_OSFILE=72	# critical OS file missing
EX_CONFIG=78	# gesogram config | environment error
EX_NOPERM=77	# permission denied


smcontrol()
{
	local whataction=$1
	local procidfile='/var/run/sendmail.pid'

	if [ "${whataction}X" = "startX" ]; then

		msg 'Starting sendmail ... '
		if [ -s "$procidfile" ]; then
			msg ", $procidfile exists"
			result NG
			STATUS_CODE=$EX_GENERAL
		else
			local mtaoptions=''
			local daemonopts="`test -s /etc/mail/daemon-options && grep -v '^#' /etc/mail/daemon-options`"
			local mqinterval="`test -s /etc/mail/queue-interval && head -1 /etc/mail/queue-interval`"

			[ -n "$SMLOG_LEVEL" ] && mtaoptions="$mtaoptions -OLogLevel=$SMLOG_LEVEL"
			[ -n "$DEBUG_LEVEL" ] && mtaoptions="$mtaoptions -d$DEBUG_LEVEL"

			[ -z "$mqinterval" ] && mqinterval='6m'
			( /usr/sbin/sendmail -bd -q$mqinterval $daemonopts $mtaoptions && sleep 1 )

			if [ "$?" -eq "0" -a -s "$procidfile" ]; then
				result OK
				STATUS_CODE=$EX_OK
			else
				result NG
				STATUS_CODE=$EX_GENERAL
			fi
		fi

	elif [ "${whataction}X" = "stopX" ]; then

		msg 'Stopping sendmail ... '
		if [ -s "$procidfile" ]; then
			( kill -$KILL_SIGNAL `head -1 $procidfile` && sleep 1 )
			if [ "$?" -eq "0" -a ! -s "$procidfile" ]; then
				result OK
				STATUS_CODE=$EX_OK
			else
				result NG
				STATUS_CODE=$EX_GENERAL
			fi
		else
			msg ", $procidfile does not exist"
			result NG
			STATUS_CODE=$EX_GENERAL
		fi
	fi
}

makemaps()
{
	for M in access genericstable virtusertable mailertable; do
		msg "makemap $M.db ... "
		if [ -f "/etc/mail/$M" ]; then
			/usr/sbin/makemap hash /etc/mail/$M.db < /etc/mail/$M > /dev/null 2>&1
			if [ "$?" -eq "0" ]; then
				result OK
				STATUS_CODE=$EX_OK
			else
				result NG
				STATUS_CODE=$EX_GENERAL
			fi
		else
			msg ', map does not exist'
			result SKIP
		fi
	done
}

makealiases()
{
	msg 'Running newaliases ... '
	if [ -f "/etc/mail/aliases" ]; then
		msg ", build /etc/mail/aliases"
		/usr/sbin/sendmail -bi > /dev/null 2>&1
		if [ "$?" -eq "0" ]; then
			chown root.$SMUSER /etc/mail/aliases.db
			chmod 0640 /etc/mail/alases.db
			result OK
			STATUS_CODE=$EX_OK
		else
			result NG
			STATUS_CODE=$EX_GENERAL
		fi
	else
		result NG
		STATUS_CODE=$EX_OSFILE
	fi
}

printqueues()
{
	queueswitch="$1"
	sendmailcmd="/usr/sbin/sendmail"
	case $queueswitch in
		L)
			$sendmailcmd -bp -qL -v
			;;

		Q|Quarantine)
			if [ "${queueswitch}X" = "QuarantineX" ]; then
				$sendmailcmd -bp -v -qQpostsend -OQueueSortOrder=none -OMaxQueueRunSize=1
			else
				$sendmailcmd -bp -v -qQpostsend -OQueueSortOrder=none -OMaxQueueRunSize=1 \
					| grep -i 'Total' | sed 's|^.*requests\?: ||i'
			fi
			;;

		q)
			$sendmailcmd -bp -OMaxQueueRunSize=1 -OQueueSortOrder=none \
				| grep -i 'Total' | sed 's|^.*requests\?: ||i'
			;;
		queue)
			$sendmailcmd -bp -OQueueSortOrder=none
			;;
	esac
}

sendmailstatus()
{
	local procidfile="/var/run/sendmail.pid"
	if [ -f "$procidfile" ]; then
		msg "sendmail is running, pid = `head -n 1 $procidfile`"
		echo > $STDERRORDEV
		msg "command args = `tail -n 1 $procidfile`"
		echo > $STDERRORDEV
	fi
}

proclist()
{
	ps axuww | grep sendmail: | grep -v grep | sed 's| *$||g'
}

msg()
{
	[ "${TERM}X" = "X" ] && STDERRORDEV='/dev/null'
	echo -n "$1" > $STDERRORDEV
}

cleandirtyqueus()
{
	# Clean up dirty queues.
	local date_expire=1
	local queue_names=''
	local queue_spool="/var/spool/queues"

	# Delete fake queues
	queue_names=`find $queue_spool -type f \
		-name "[A-Z]f??????????????" -exec basename {} \; | sed 's|^..||g' | sort | uniq`

	if [ -n "$queue_names" ]; then
		for qn in $queue_names; do
			find $queue_spool -type f -name "[Qdlntxq]f$qn" -delete
		done
	fi

	# Delete expired queues
	find $queue_spool -type f -atime +$date_expire -name "[A-Za-z]f??????????????" -delete
}

result()
{
	local msgtype="$1"
	local message=''
	[ "${TERM}X" = "X" ] && STDERRORDEV='/dev/null'

	case "$msgtype" in
		ok|OK)
			message=$CHAR_OK
			;;

		ng|NG)
			message=$CHAR_NG
			;;

		warn|WARN)
			message=$CHAR_WARN
			;;

		info|INFO)
			message=$CHAR_INFO
			;;

		skip|SKIP)
			message=$CHAR_SKIP
			;;
		*)
			;;
	esac

	printf \\033[70G > $STDERRORDEV
	printf \\033[1\;49m > $STDERRORDEV
	echo -n $message > $STDERRORDEV
	printf \\033[0\;39m > $STDERRORDEV
	echo > $STDERRORDEV
}

init()
{
	export PATH='/bin:/usr/bin:/sbin:/usr/sbin:/usr/local/bin:/usr/local/sbin'
	export LANG='C'
	if [ -x "/usr/sbin/sendmail" ]; then 
		:
	else
		echo "Cannot execute /usr/sbin/sendmail" > $STDERRORDEV
		exit $EX_GENERAL
	fi
}

help()
{
	echo
	echo 'usage:'
	echo " $MYNAME [ OPTION [ARGV] ] COMMAND"
	echo
	echo '  COMMAND:'
	echo '     start        : Start sendmail daemon.'
	echo '     stop         : Stop sendmail.'
	echo '     restart      : Restart ( stop & start )'
	echo '     status,s     : Print sendmail status, content of pid file.'
	echo '     mailq        : Print the detail of mail queue.'
	echo '     newaliases   : Run newaliases command.'
	echo '     queue,q      : Print the number of mail queues.'
	echo '     ps           : Print sendmail processes.'
	echo '     clean        : Clean up dirty queues and dirty processes( sendmail should be stopped ).'
	echo '     help,h       : Print this screen. '
	echo 

	echo '  OPTIONS:'
	echo '   * boot options:'
	echo '     -d <n|KEY>   : Enable debug mode (-d<n>) and output transwhataction logs'
	echo '     -L <n>       : Specify LogLevel'
	echo '     -M           : Run makemap before starting.'
	echo "     -v           : Print $MYNAME/sendmail version"
	echo

	if [ $ENABLE_DEBUG -eq 1 ]; then
		echo '  KEY: ( debug mode )'
		echo '    agent alias check class database error header io'
		echo '    macro mailer mci option queue ruleset syslog '
		echo
		echo '  ENVIRONMENT:'
		echo '   sendmail-8.14.0 or later required'
		echo
	fi
}

init
while getopts ":d:L:Mv" opt; do
	case $opt in
		d)
			# Enable debug mode
			case "$OPTARG" in
				agent|mailer)
					DEBUG_LEVEL='11.2'
					;;
				alias)
					DEBUG_LEVEL='27.1'
					;;
				check)
					DEBUG_LEVEL='48.2'
					;;
				class)
					DEBUG_LEVEL='37.8'
					;;
				database)
					DEBUG_LEVEL='38.20'
					;;
				error)
					DEBUG_LEVEL='6.1'
					;;
				header)
					DEBUG_LEVEL='31.10'
					;;
				io)
					DEBUG_LEVEL='58.20'
					;;
				macro)
					DEBUG_LEVEL='35.13'
					;;
				mci)
					DEBUG_LEVEL='42.20'
					;;
				option)
					DEBUG_LEVEL='37.1'
					;;
				quarantine)
					DEBUG_LEVEL='70.1'
					;;
				queue)
					DEBUG_LEVEL='41.1'
					;;
				ruleset)
					DEBUG_LEVEL='21.12'
					;;
				syslog)
					DEBUG_LEVEL='0.0'
					;;
				*)
					DEBUG_LEVEL="$OPTARG"
					;;
			esac

			ENABLE_DEBUG=1
			;;

		L)
			# Specify LogLevel option(external only)
			[ -n "$OPTARG" ] && SMLOG_LEVEL="$OPTARG"
			;;
		M)
			# Run makemap command when boot
			RUN_MAKEMAP=1
			;;
		v)
			# Print my version.
			SM_VERSION="`/usr/sbin/sendmail -d0.101 root < /dev/null 2> /dev/null | awk '{print $2}'`"
			echo -n "$SYSVERSION"
			test -n "$SM_VERSION" && echo -n "/$SM_VERSION"
			echo
			exit $EX_OK
			;;
		\?|*)
			;;
	esac
done

shift $( expr $OPTIND - 1 )


#  ____ ____ ____ ____ 
# ||m |||a |||i |||n ||
# ||__|||__|||__|||__||
# |/__`|/__`|/__`|/__`|
#
case "$1" in
	#  ____ ____ ____ _________ ____ ____ ____ ____ ____ ____ ____ ____ 
	# ||s |||u |||b |||       |||c |||o |||m |||m |||a |||n |||d |||s ||
	# ||__|||__|||__|||_______|||__|||__|||__|||__|||__|||__|||__|||__||
	# |/__`|/__`|/__`|/_______`|/__`|/__`|/__`|/__`|/__`|/__`|/__`|/__`|
	#
	clean)
		cleandirtyqueus
		;;

	help|h)
		help
		exit $EX_OK
		;;

	mailq)
		# Print the number of mail queues.
		# Compatible with the sendmail 8.11 ( no queue groups )
		printqueues mailq
		exit $EX_OK
		;;

	ps)
		# Print processes.
		proclist
		exit $EX_OK
		;;

	queue|q)
		# Print the number of mail queues.
		printqueues $1
		exit $EX_OK
		;;
	
	newaliases)
		# sendmail -bi
		makealiases
		exit $STATUS_CODE
		;;

	restart)
		smcontrol stop
		[ "$RUN_MAKEMAP" -eq "1" ] && makemaps
		smcontrol start
		;;

	start)
		[ "$RUN_MAKEMAP" -eq "1" ] && makemaps
		smcontrol start
		;;

	status|s)
		sendmailstatus
		;;

	stop|abort)
		if [ "$1" = "abort" ]; then
			KILL_SIGNAL=9	
		fi
		smcontrol stop
		;;
	
	*)
		echo "try $MYNAME help"
		exit $EX_USAGE
		;;
esac

